File: backend/src/test/java/com/worklog/domain/attendance/DailyAttendanceTest.java
package com.worklog.domain.attendance;

import static org.junit.jupiter.api.Assertions.*;

import com.worklog.domain.member.MemberId;
import com.worklog.domain.tenant.TenantId;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.UUID;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

@DisplayName("DailyAttendance entity")
class DailyAttendanceTest {

    private static final TenantId TENANT = TenantId.of("550e8400-e29b-41d4-a716-446655440001");
    private static final MemberId MEMBER = MemberId.of("660e8400-e29b-41d4-a716-446655440001");
    private static final LocalDate DATE = LocalDate.of(2026, 3, 2);

    @Test
    @DisplayName("create sets defaults")
    void create_setsDefaults() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);

        assertNotNull(attendance.getId());
        assertEquals(TENANT, attendance.getTenantId());
        assertEquals(MEMBER, attendance.getMemberId());
        assertEquals(DATE, attendance.getDate());
        assertNull(attendance.getStartTime());
        assertNull(attendance.getEndTime());
        assertNull(attendance.getRemarks());
        assertEquals(0, attendance.getVersion());
    }

    @Test
    @DisplayName("update changes mutable fields")
    void update_changesMutableFields() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);

        attendance.update(LocalTime.of(9, 0), LocalTime.of(18, 0), "Meeting day");

        assertEquals(LocalTime.of(9, 0), attendance.getStartTime());
        assertEquals(LocalTime.of(18, 0), attendance.getEndTime());
        assertEquals("Meeting day", attendance.getRemarks());
    }

    @Test
    @DisplayName("update rejects end before start")
    void update_rejectsEndBeforeStart() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);

        assertThrows(
                IllegalArgumentException.class, () -> attendance.update(LocalTime.of(18, 0), LocalTime.of(9, 0), null));
    }

    @Test
    @DisplayName("update allows start time only")
    void update_allowsStartTimeOnly() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);

        attendance.update(LocalTime.of(9, 0), null, null);

        assertEquals(LocalTime.of(9, 0), attendance.getStartTime());
        assertNull(attendance.getEndTime());
    }

    @Test
    @DisplayName("update rejects long remarks")
    void update_rejectsLongRemarks() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
        String longRemarks = "a".repeat(501);

        assertThrows(
                IllegalArgumentException.class,
                () -> attendance.update(LocalTime.of(9, 0), LocalTime.of(18, 0), longRemarks));
    }

    @Test
    @DisplayName("update allows 500 char remarks")
    void update_allows500CharRemarks() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
        String remarks500 = "a".repeat(500);

        attendance.update(LocalTime.of(9, 0), LocalTime.of(18, 0), remarks500);

        assertEquals(remarks500, attendance.getRemarks());
    }

    @Test
    @DisplayName("update allows nulls")
    void update_allowsNulls() {
        DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
        attendance.update(LocalTime.of(9, 0), LocalTime.of(18, 0), "Meeting day");

        attendance.update(null, null, null);

        assertNull(attendance.getStartTime());
        assertNull(attendance.getEndTime());
        assertNull(attendance.getRemarks());
    }

    @Nested
    @DisplayName("Equality and HashCode")
    class EqualityAndHashCode {

        @Test
        @DisplayName("should be equal to itself")
        void equalToItself() {
            DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
            assertEquals(attendance, attendance);
        }

        @Test
        @DisplayName("should be equal when same ID")
        void equalWhenSameId() {
            UUID sharedUuid = UUID.randomUUID();
            DailyAttendance a1 =
                    new DailyAttendance(DailyAttendanceId.of(sharedUuid), TENANT, MEMBER, DATE, null, null, null, 0);
            DailyAttendance a2 = new DailyAttendance(
                    DailyAttendanceId.of(sharedUuid),
                    TENANT,
                    MEMBER,
                    DATE,
                    LocalTime.of(9, 0),
                    LocalTime.of(18, 0),
                    "different fields",
                    1);
            assertEquals(a1, a2);
        }

        @Test
        @DisplayName("should not be equal when different ID")
        void notEqualWhenDifferentId() {
            DailyAttendance a1 = new DailyAttendance(
                    DailyAttendanceId.of(UUID.randomUUID()), TENANT, MEMBER, DATE, null, null, null, 0);
            DailyAttendance a2 = new DailyAttendance(
                    DailyAttendanceId.of(UUID.randomUUID()), TENANT, MEMBER, DATE, null, null, null, 0);
            assertNotEquals(a1, a2);
        }

        @Test
        @DisplayName("should not be equal to null")
        void notEqualToNull() {
            DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
            assertFalse(attendance.equals(null));
        }

        @Test
        @DisplayName("should not be equal to different type")
        void notEqualToDifferentType() {
            DailyAttendance attendance = DailyAttendance.create(TENANT, MEMBER, DATE);
            assertFalse(attendance.equals("not an attendance"));
        }

        @Test
        @DisplayName("should have consistent hashCode when same ID")
        void hashCodeConsistentWithEquals() {
            UUID sharedUuid = UUID.randomUUID();
            DailyAttendance a1 =
                    new DailyAttendance(DailyAttendanceId.of(sharedUuid), TENANT, MEMBER, DATE, null, null, null, 0);
            DailyAttendance a2 =
                    new DailyAttendance(DailyAttendanceId.of(sharedUuid), TENANT, MEMBER, DATE, null, null, null, 0);
            assertEquals(a1.hashCode(), a2.hashCode());
        }
    }
}

File: backend/src/test/kotlin/com/worklog/domain/attendance/DailyAttendanceTest.kt
package com.worklog.domain.attendance

import com.worklog.domain.member.MemberId
import com.worklog.domain.tenant.TenantId
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.LocalTime
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for DailyAttendance entity.
 */
class DailyAttendanceTest {
    private val tenantId = TenantId(UUID.randomUUID())
    private val memberId = MemberId(UUID.randomUUID())
    private val attendanceDate = LocalDate.of(2025, 6, 15)

    @Test
    fun `create should generate id and set defaults`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                LocalTime.of(18, 0),
                "Normal day",
            )

        assertNotNull(attendance.id)
        assertEquals(tenantId, attendance.tenantId)
        assertEquals(memberId, attendance.memberId)
        assertEquals(attendanceDate, attendance.attendanceDate)
        assertEquals(LocalTime.of(9, 0), attendance.startTime)
        assertEquals(LocalTime.of(18, 0), attendance.endTime)
        assertEquals("Normal day", attendance.remarks)
        assertEquals(0, attendance.version)
    }

    @Test
    fun `create should allow null times and remarks`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                null,
                null,
                null,
            )

        assertNotNull(attendance.id)
        assertEquals(tenantId, attendance.tenantId)
        assertEquals(memberId, attendance.memberId)
        assertEquals(attendanceDate, attendance.attendanceDate)
        assertNull(attendance.startTime)
        assertNull(attendance.endTime)
        assertNull(attendance.remarks)
        assertEquals(0, attendance.version)
    }

    @Test
    fun `should reject endTime before startTime`() {
        val exception =
            assertFailsWith<IllegalArgumentException> {
                DailyAttendance.create(
                    tenantId,
                    memberId,
                    attendanceDate,
                    LocalTime.of(18, 0),
                    LocalTime.of(9, 0),
                    null,
                )
            }

        assertEquals("End time cannot be before start time", exception.message)
    }

    @Test
    fun `should reject remarks over 500 characters`() {
        val longRemarks = "A".repeat(501)

        val exception =
            assertFailsWith<IllegalArgumentException> {
                DailyAttendance.create(
                    tenantId,
                    memberId,
                    attendanceDate,
                    null,
                    null,
                    longRemarks,
                )
            }

        assertEquals("Remarks cannot exceed 500 characters", exception.message)
    }

    @Test
    fun `should allow equal startTime and endTime`() {
        val time = LocalTime.of(12, 0)

        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                time,
                time,
                null,
            )

        assertEquals(time, attendance.startTime)
        assertEquals(time, attendance.endTime)
    }

    @Test
    fun `should allow only startTime set`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                null,
                null,
            )

        assertEquals(LocalTime.of(9, 0), attendance.startTime)
        assertNull(attendance.endTime)
    }

    @Test
    fun `should allow only endTime set`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                null,
                LocalTime.of(18, 0),
                null,
            )

        assertNull(attendance.startTime)
        assertEquals(LocalTime.of(18, 0), attendance.endTime)
    }

    @Test
    fun `should allow remarks exactly 500 characters`() {
        val remarks = "A".repeat(500)

        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                null,
                null,
                remarks,
            )

        assertEquals(500, attendance.remarks.length)
    }

    @Test
    fun `should allow empty remarks`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                null,
                null,
                "",
            )

        assertEquals("", attendance.remarks)
    }

    @Test
    fun `constructor should reject null tenantId`() {
        assertFailsWith<NullPointerException> {
            DailyAttendance(
                DailyAttendanceId.generate(),
                null,
                memberId,
                attendanceDate,
                null,
                null,
                null,
                0,
            )
        }
    }

    @Test
    fun `constructor should reject null memberId`() {
        assertFailsWith<NullPointerException> {
            DailyAttendance(
                DailyAttendanceId.generate(),
                tenantId,
                null,
                attendanceDate,
                null,
                null,
                null,
                0,
            )
        }
    }

    @Test
    fun `constructor should reject null attendanceDate`() {
        assertFailsWith<NullPointerException> {
            DailyAttendance(
                DailyAttendanceId.generate(),
                tenantId,
                memberId,
                null,
                null,
                null,
                null,
                0,
            )
        }
    }

    @Test
    fun `update should change mutable fields`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                LocalTime.of(18, 0),
                "Original",
            )

        attendance.update(LocalTime.of(10, 0), LocalTime.of(19, 0), "Updated")

        assertEquals(LocalTime.of(10, 0), attendance.startTime)
        assertEquals(LocalTime.of(19, 0), attendance.endTime)
        assertEquals("Updated", attendance.remarks)
    }

    @Test
    fun `update should allow clearing times`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                LocalTime.of(18, 0),
                "note",
            )

        attendance.update(null, null, null)

        assertNull(attendance.startTime)
        assertNull(attendance.endTime)
        assertNull(attendance.remarks)
    }

    @Test
    fun `update should reject invalid time range`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                LocalTime.of(18, 0),
                null,
            )

        assertFailsWith<IllegalArgumentException> {
            attendance.update(LocalTime.of(18, 0), LocalTime.of(9, 0), null)
        }
    }

    @Test
    fun `update should reject long remarks`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                null,
                null,
                null,
            )

        assertFailsWith<IllegalArgumentException> {
            attendance.update(null, null, "A".repeat(501))
        }
    }

    @Test
    fun `equals should be true for same id`() {
        val id = DailyAttendanceId.generate()
        val a1 = DailyAttendance(id, tenantId, memberId, attendanceDate, null, null, null, 0)
        val a2 = DailyAttendance(id, tenantId, memberId, attendanceDate, LocalTime.of(9, 0), null, "note", 1)

        assertEquals(a1, a2)
    }

    @Test
    fun `equals should be false for different id`() {
        val a1 =
            DailyAttendance(DailyAttendanceId.generate(), tenantId, memberId, attendanceDate, null, null, null, 0)
        val a2 =
            DailyAttendance(DailyAttendanceId.generate(), tenantId, memberId, attendanceDate, null, null, null, 0)

        assertNotEquals(a1, a2)
    }

    @Test
    fun `equals should be false for null`() {
        val a = DailyAttendance.create(tenantId, memberId, attendanceDate, null, null, null)

        assertNotEquals(null, a)
    }

    @Test
    fun `equals should be false for different type`() {
        val a = DailyAttendance.create(tenantId, memberId, attendanceDate, null, null, null)

        assertFalse(a.equals("not an attendance"))
    }

    @Test
    fun `hashCode should be consistent for same id`() {
        val id = DailyAttendanceId.generate()
        val a1 = DailyAttendance(id, tenantId, memberId, attendanceDate, null, null, null, 0)
        val a2 = DailyAttendance(id, tenantId, memberId, attendanceDate, null, null, null, 1)

        assertEquals(a1.hashCode(), a2.hashCode())
    }

    @Test
    fun `toString should contain key fields`() {
        val attendance =
            DailyAttendance.create(
                tenantId,
                memberId,
                attendanceDate,
                LocalTime.of(9, 0),
                LocalTime.of(18, 0),
                null,
            )
        val str = attendance.toString()

        assertTrue(str.contains("DailyAttendance"))
        assertTrue(str.contains(memberId.toString()))
        assertTrue(str.contains(attendanceDate.toString()))
    }
}
